src/drag_icon.c

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: GPL-2.0-only
+#include <assert.h>
+#include <wlr/types/wlr_data_device.h>
+#include <wlr/types/wlr_scene.h>
+#include <wlr/util/log.h>
+#include "drag_icon.h"
+#include "labwc.h"  /* for struct seat */
+
+static void
+on_icon_map(struct wl_listener *listener, void *data)
+{
+	struct drag_icon *self = wl_container_of(listener, self, events.map);
+	struct wlr_drag_icon *icon = data;
+	icon->data = wlr_scene_subsurface_tree_create(self->tree, icon->surface);
+	wlr_log(WLR_ERROR, "Drag icon mapped");
+}
+
+static void
+on_surface_commit(struct wl_listener *listener, void *data)
+{
+	struct drag_icon *self = wl_container_of(listener, self, events.commit);
+	struct wlr_surface *surface = data;
+	struct wlr_scene_tree *surface_tree = self->icon->data;
+	wlr_scene_node_set_position(&surface_tree->node, surface->sx, surface->sy);
+	wlr_log(WLR_ERROR, "Drag surface committed to pos %d,%d", surface->sx, surface->sy);
+}
+
+static void
+on_icon_unmap(struct wl_listener *listener, void *data)
+{
+	struct drag_icon *self = wl_container_of(listener, self, events.unmap);
+	struct wlr_drag_icon *icon = data;
+	wlr_scene_node_set_enabled(icon->data, false);
+	wlr_log(WLR_ERROR, "Drag icon unmapped");
+}
+
+static void
+on_icon_destroy(struct wl_listener *listener, void *data)
+{
+	struct drag_icon *self = wl_container_of(listener, self, events.destroy);
+	wlr_scene_node_destroy(&self->tree->node);
+
+	wl_list_remove(&self->events.map.link);
+	wl_list_remove(&self->events.unmap.link);
+	wl_list_remove(&self->events.destroy.link);
+	wl_list_remove(&self->events.commit.link);
+	wl_list_remove(&self->link);
+	free(self);
+	wlr_log(WLR_ERROR, "Drag icon destroyed");
+}
+
+void
+drag_icon_create(struct seat *seat, struct wlr_drag_icon *wlr_icon)
+{
+	struct drag_icon *self = calloc(1, sizeof(*self));
+
+	/* Set up parent tree */
+	self->icon = wlr_icon;
+	self->tree = wlr_scene_tree_create(&seat->server->scene->tree);
```

Comment:
Potentially create a shared sub tree for all icons instead so we don't have to loop through all of them and instead just position the parent tree based on the cursor coordinates. Same for showing / hiding to prevent being on top when doing hitbox detection.

Reply:
`seat->drag.icons` is now the shared parent scene tree for all drag icons.